airflow/models/dagcode.py

```diff
@@ -0,0 +1,147 @@
+from typing import Iterable, List
+
+from sqlalchemy import BigInteger, Column, String, UnicodeText, and_, exists
+
+from airflow.exceptions import AirflowException
+from airflow.models import Base
+from airflow.utils import timezone
+from airflow.utils.file import correct_maybe_zipped, open_maybe_zipped
+from airflow.utils.session import provide_session
+from airflow.utils.sqlalchemy import UtcDateTime
+
+log = logging.getLogger(__name__)
+
+
+class DagCode(Base):
+    """A table for DAGs code.
+
+    dag_code table contains code of DAG files synchronized by scheduler.
+    This feature is controlled by:
+
+    * ``[core] store_serialized_dags = True``: enable this feature
+    * ``[core] store_dag_code = True``: enable this feature
+
+    For details on dag serialization see SerializedDagModel
+    """
+    __tablename__ = 'dag_code'
+
+    fileloc_hash = Column(
+        BigInteger, nullable=False, primary_key=True, autoincrement=False)
+    fileloc = Column(String(2000), nullable=False)
+    # The max length of fileloc exceeds the limit of indexing.
+    last_updated = Column(UtcDateTime, nullable=False)
+    source_code = Column(UnicodeText(), nullable=False)
+
+    def __init__(self, full_filepath: str):
+        self.fileloc = full_filepath
+        self.fileloc_hash = DagCode.dag_fileloc_hash(self.fileloc)
+        self.last_updated = timezone.utcnow()
+        self.source_code = DagCode._read_code(self.fileloc)
+
+    @classmethod
+    def _read_code(cls, fileloc: str):
+        with open_maybe_zipped(fileloc, 'r') as source:
+            source_code = source.read()
+        return source_code
+
+    @provide_session
+    def sync_to_db(self, session=None):
+        """Writes code into database.
+
+        :param session: ORM Session
+        """
+        old_version = session.query(
+            DagCode.fileloc, DagCode.fileloc_hash, DagCode.last_updated) \
+            .filter(DagCode.fileloc_hash == self.fileloc_hash) \
+            .first()
+
+        if old_version and old_version.fileloc != self.fileloc:
+            raise AirflowException(
+                "Filename '{}' causes a hash collision in the database with "
+                "'{}'. Please rename the file.".format(
+                    self.fileloc, old_version.fileloc))
+
+        file_modified = datetime.fromtimestamp(
+            os.path.getmtime(correct_maybe_zipped(self.fileloc)), tz=timezone.utc)
+
+        if old_version and (file_modified - timedelta(seconds=120)) < \
+                old_version.last_updated:
+            return
+
+        session.merge(self)
+
+    @classmethod
+    @provide_session
+    def bulk_sync_to_db(cls, filelocs: Iterable[str], session=None):
+        """Writes code in bulk into database.
+
+        :param filelocs: file paths of DAGs to sync
+        :param session: ORM Session
+        """
+        filelocs = set(filelocs)
+        for file in filelocs:
+            DagCode(file).sync_to_db(session=session)
```

Comment:
I believe if it is adding/updating session and inside transaction it is not committed until somebody calls session.commit().
But I am new to Python and SqlAlchemy.
But, anyway, I have modified the implementation, to do only one query for old records instead of multiple queries by fileloc.

Reply:
Resolve if you believe new batch update/insert implementation is correct.